lib/quicrq.c

```diff
@@ -1788,6 +1789,101 @@ int quicrq_receive_stream_data(quicrq_stream_ctx_t* stream_ctx, uint8_t* bytes,
+                                quicrq_stream_ctx_t* ctrl_stream_ctx = quicrq_get_control_stream_for_media_id(cnx_ctx, incoming.media_id);
+                                uni_stream_ctx->receive_state = quicrq_receive_warp_header;
+                                uni_stream_ctx->control_stream_ctx = ctrl_stream_ctx;
+                                /* todo: add check for null control stream */
+                                quicrq_log_message(cnx_ctx, "UniStream %" PRIu64 ", received warp header message uni_id= %" PRIu64,
+                                                   ", ControlStream id= %" PRIu64 " media id=%" PRIu64,
+                                                   uni_stream_ctx->stream_id, uni_stream_ctx->control_stream_ctx->stream_id, incoming.media_id);
+
+                                break;
+                            case QUICRQ_ACTION_OBJECT_HEADER:
+                                if (uni_stream_ctx->receive_state != quicrq_receive_warp_header) {
+                                    /* Protocol error */
+                                    ret = -1;
+                                }
+                                else {
+                                    uni_stream_ctx->receive_state = quicrq_receive_object_header;
+                                    quicrq_stream_ctx_t* ctrl_stream_ctx = quicrq_get_control_stream_for_media_id(cnx_ctx, incoming.media_id);
+                                    uni_stream_ctx->control_stream_ctx = ctrl_stream_ctx;
+                                    /* Pass the repair data to the media consumer. */
+                                    ret = ctrl_stream_ctx->consumer_fn(quicrq_media_datagram_ready, ctrl_stream_ctx->media_ctx, picoquic_get_quic_time(ctrl_stream_ctx->cnx_ctx->qr_ctx->quic),
+                                                                  incoming.data, uni_stream_ctx->current_group_id, incoming.object_id,
+                                                                  incoming.offset, 0, incoming.flags, incoming.nb_objects_previous_group,
+                                                                  incoming.is_last_fragment, incoming.length);
+                                    ret = quicrq_cnx_handle_consumer_finished(ctrl_stream_ctx, 0, 0, ret);
+                                }
+                                break;
+                                default:
+                                    /* Some unknown message, maybe not implemented yet */
+                                    ret = -1;
+                                    break;
+                        }
+                    }
+                    /* As the message was processed, reset the message buffer. */
+                    quicrq_msg_buffer_reset(&uni_stream_ctx->message_buffer);
+                }
+            }
+        }
+    }
+
+    if (is_fin) {
+        /* TODO: What needs to be done here */
+    }
```

Comment:
Close the stream, that's what needs to be done. But only after all bytes have been processed....